src/main/java/htsjdk/samtools/util/htsget/SAMRecordPrefetchingIterator.java

```diff
@@ -0,0 +1,85 @@
+package htsjdk.samtools.util.htsget;
+
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.util.CloseableIterator;
+import htsjdk.samtools.util.PeekableIterator;
+
+import java.util.NoSuchElementException;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class SAMRecordPrefetchingIterator implements CloseableIterator<SAMRecord> {
```

Comment:
I think this class could use specific tests for it.  I think we could write tests that use a mocked input iterator that allows us to test various cases, (things like the reading throwing exceptions), very large reads compared to the number of tickets, etc